pallets/dca/src/lib.rs

```diff
@@ -0,0 +1,971 @@
+		) -> DispatchResult {
+			let who = ensure_signed(origin.clone())?;
+			ensure!(who == schedule.owner, Error::<T>::Forbidden);
+
+			ensure!(schedule.order.get_route_length() > 0, Error::<T>::RouteNotSpecified);
+
+			let min_budget = Self::convert_native_amount_to_currency(
+				schedule.order.get_asset_in(),
+				T::MinBudgetInNativeCurrency::get(),
+			)?;
+			ensure!(
+				schedule.total_amount >= min_budget,
+				Error::<T>::TotalAmountIsSmallerThanMinBudget
+			);
+
+			let transaction_fee = Self::get_transaction_fee(&schedule.order)?;
+
+			let amount_in = match schedule.order {
+				Order::Sell { amount_in, .. } => amount_in,
+				Order::Buy {
+					amount_out, ref route, ..
+				} => Self::get_amount_in_for_buy(&amount_out, route)?,
+			};
+			ensure!(amount_in > transaction_fee, Error::<T>::TradeAmountIsLessThanFee);
+
+			let amount_in_with_transaction_fee = amount_in
+				.checked_add(transaction_fee)
+				.ok_or(ArithmeticError::Overflow)?;
+			ensure!(
+				amount_in_with_transaction_fee <= schedule.total_amount,
+				Error::<T>::BudgetTooLow
+			);
+
+			let next_schedule_id =
+				ScheduleIdSequencer::<T>::try_mutate(|current_id| -> Result<ScheduleId, DispatchError> {
+					let schedule_id = *current_id;
+					*current_id = current_id.checked_add(1).ok_or(ArithmeticError::Overflow)?;
+					Ok(schedule_id)
+				})?;
+
+			Schedules::<T>::insert(next_schedule_id, &schedule);
+			ScheduleOwnership::<T>::insert(who.clone(), next_schedule_id, ());
+			RemainingAmounts::<T>::insert(next_schedule_id, schedule.total_amount);
+			RetriesOnError::<T>::insert(next_schedule_id, 0);
+
+			T::Currencies::reserve_named(
+				&T::NamedReserveId::get(),
+				schedule.order.get_asset_in(),
+				&who,
+				schedule.total_amount,
+			)?;
+
+			let blocknumber_for_first_schedule_execution = match start_execution_block {
+				Some(blocknumber) => Ok(blocknumber),
+				None => {
+					let current_block_number = frame_system::Pallet::<T>::current_block_number();
+					let next_block_number = current_block_number
+						.checked_add(&T::BlockNumber::one())
+						.ok_or(ArithmeticError::Overflow)?;
+
+					Ok::<T::BlockNumber, ArithmeticError>(next_block_number)
+				}
+			}?;
+
+			Self::plan_schedule_for_block(&who, blocknumber_for_first_schedule_execution, next_schedule_id)?;
+
+			Self::deposit_event(Event::Scheduled {
+				id: next_schedule_id,
+				who,
+			});
+
+			Ok(())
+		}
+
+		/// Admin endpoint to terminate a DCA schedule and remove it completely from the chain.
+		///
+		/// Parameters:
+		/// - `origin`: schedule owner
+		/// - `schedule_id`: schedule id
+		/// - `next_execution_block`: block number where the schedule is planned.
+		///
+		/// Emits `Terminated` event when successful.
+		#[pallet::call_index(1)]
+		#[pallet::weight(<T as Config>::WeightInfo::terminate())]
+		#[transactional]
+		pub fn terminate(
+			origin: OriginFor<T>,
+			schedule_id: ScheduleId,
+			next_execution_block: BlockNumberFor<T>,
```

Comment:
this would be very hard to fill if you have schedule with period of 1 (every block), i would either allow `None` here that would then filled next block automatically, this would however only solve period of 1, but period of 2 e.g. could be still troublesome to terminate however. alternatively i would disallow schedules with period less than something (10 blocks for example)